### ISSUE TITLE
Sidecar fails to update Nginx with Go 1.7.1

### ISSUE THREAD

**Issue by zcahana** [amalgam8/amalgam8#266]
Following the upgrade to Go 1.7.1 (#263), I'm seeing the following errors in the gateway's logs when running the integration tests:

```
time="2016-09-19T09:36:16Z" level=error msg="Failed to send request to NGINX server" error="Post http://localhost:5813/a8-admin: dial tcp: i/o timeout" 
time="2016-09-19T09:36:16Z" level=error msg="Failed to update HTTP upstreams with NGINX" error="Post http://localhost:5813/a8-admin: dial tcp: i/o timeout" 
time="2016-09-19T09:36:16Z" level=warning msg="Controller listener failed" error="Post http://localhost:5813/a8-admin: dial tcp: i/o timeout" 
time="2016-09-19T09:36:46Z" level=error msg="Failed to send request to NGINX server" error="Post http://localhost:5813/a8-admin: dial tcp: i/o timeout" 
time="2016-09-19T09:36:46Z" level=error msg="Failed to update HTTP upstreams with NGINX" error="Post http://localhost:5813/a8-admin: dial tcp: i/o timeout" 
time="2016-09-19T09:36:46Z" level=warning msg="Registry listener failed" error="Post http://localhost:5813/a8-admin: dial tcp: i/o timeout" 
```

... which shows that the sidecar isn't able to properly set up a TCP connection with the Lua server running within Nginx on port 5813. Manually exec'ing into the gateway container, I can successfully run `curl localhost:5813/a8-admin` repeatedly.

Note that I've tested this multiple times with Go 1.6.2, 1.7 and 1.7.1. This is repeatedly happening with Go 1.7.1, but not on 1.6.2 nor 1.7, so circumstantially I'm tracing this to Go 1.7.1.

**Comment by zcahana** [comment]
Further investigation shows that this is related to DNS resolution of "localhost" (as part of `POST http://localhost:5813`) within the sidecar container. Note that this happens only when running within K8s:
- On Go 1.7.1, it just times out after 30s (hence the `dial tcp: i/o timeout` above)
- On Go 1.7, it succeeds, but DNS resolution takes 20s.
- On Go <= 1.6.3, it succeeds immediately.

This seems to be related to golang/go#16865, which supposedly fixed a regression in Go 1.7 around DNS resolution logic.

**Comment by zcahana** [comment]
@ericvn Note that we've temporarily reverted back to Go 1.6.3, until the issue is resolved.

**Comment by ericvn** [comment]
No problem. 1.6.3 covers the issue with Golang suppport with the new Mac OS X being deployed tomorrow which was one of the reasons to move up.

**Comment by zcahana** [comment]
This was caused by an invalid DNS setup for the Kubernetes cluster, combined with a change in the behavior of the Go-based DNS resolver in Go 1.7:

Our K8s installation scripts (in [examples/](https://github.com/amalgam8/amalgam8/blob/d3693d48f914c87c43ed44f2dc0bfa0a171b35b7/examples/kubernetes/install-kubernetes.sh#L43) and [testing/](https://github.com/amalgam8/amalgam8/blob/d3693d48f914c87c43ed44f2dc0bfa0a171b35b7/testing/kubernetes/install-kubernetes.sh#L43)) configured nodes/pods with a DNS server at `10.0.0.10`, but didn't actually start a DNS server at that address.

Also, Go 1.7 has changed the (complex) logic which defines whether to first look at `/etc/hosts`, or to query DNS servers listed in `/etc/resolv.conf`. This depends on OS, contents of `/etc/nsswitch.conf`, static linking, etc...
- On Go <= 1.6.3, `/etc/hosts` was inspected first, so `localhost` was successfully & immediatly resolved to `127.0.0.1`.
- On Go 1.7, `10.0.0.10` was queried first, timed-out after 20s, and then fallback'ed to `/etc/hosts`.
- On Go 1.7.1, the DNS lookup timeout was extended back to 40s (as in Go <= 1.6.3), but the HTTP request timeout defaulted to 30s, so the HTTP request failed before DNS fallback'ed to `/etc/hosts`.
